File: cpp/hardware/expansion/NetworkCard.cpp
#pragma once
#include <bits/stdc++.h>

#include "ExpansionCard.cpp"

class NetworkCard : public ExpansionCard
{
private:
    std::string macAddress;
    int maxSpeedMbps;
    bool supportsWifi;
    bool supportsBluetooth;

public:
    // Constructors
    NetworkCard() : ExpansionCard(), macAddress(""), maxSpeedMbps(0), supportsWifi(false), supportsBluetooth(false) {}
    NetworkCard(const std::string &type, const std::string &serial, const Date &date, bool installed, const std::string &id, const std::string &manufacturer, const std::string &model, double price, std::string mac, int speed, bool wifi, bool bluetooth)
        : ExpansionCard(type, serial, date, installed, id, manufacturer, model, price), macAddress(mac), maxSpeedMbps(speed), supportsWifi(wifi), supportsBluetooth(bluetooth) {}

    // Destructor
    ~NetworkCard() {}

    // Setters
    void setMacAddress(const std::string &mac)
    {
        macAddress = mac;
    }

    void setMaxSpeedMbps(int speed)
    {
        maxSpeedMbps = speed;
    }

    void setSupportsWifi(bool wifi)
    {
        supportsWifi = wifi;
    }

    void setSupportsBluetooth(bool bluetooth)
    {
        supportsBluetooth = bluetooth;
    }

    // Getters
    std::string getMacAddress() const
    {
        return macAddress;
    }

    int getMaxSpeedMbps() const
    {
        return maxSpeedMbps;
    }

    bool isWifiSupported() const
    {
        return supportsWifi;
    }

    bool isBluetoothSupported() const
    {
        return supportsBluetooth;
    }

    bool connect() override
    {
        printf("Network card connected\n");
        return true;
    }

    bool disconnect() override
    {
        printf("Network card disconnected\n");
        return true;
    }
};

File: python/hardware/expansion/network_card.py
from datetime import date
from hardware.expansion.expansion_card import ExpansionCard

class NetworkCard(ExpansionCard):
    def __init__(self, interface_type="", serial_number="", installation_date=None, physically_installed=False, id="", manufacturer="", model="", price=0.0, mac_address="", max_speed_mbps=0, supports_wifi=False, supports_bluetooth=False):
        super().__init__(interface_type, serial_number, installation_date, physically_installed, id, manufacturer, model, price)
        self._mac_address = mac_address
        self._max_speed_mbps = max_speed_mbps
        self._supports_wifi = supports_wifi
        self._supports_bluetooth = supports_bluetooth

    # Setters
    def set_mac_address(self, mac_address):
        self._mac_address = mac_address

    def set_max_speed_mbps(self, max_speed_mbps):
        self._max_speed_mbps = max_speed_mbps

    def set_supports_wifi(self, supports_wifi):
        self._supports_wifi = supports_wifi

    def set_supports_bluetooth(self, supports_bluetooth):
        self._supports_bluetooth = supports_bluetooth

    # Getters
    def get_mac_address(self):
        return self._mac_address

    def get_max_speed_mbps(self):
        return self._max_speed_mbps

    def is_wifi_supported(self):
        return self._supports_wifi

    def is_bluetooth_supported(self):
        return self._supports_bluetooth

    def connect(self):
        print("Network card connected")
        return True

    def disconnect(self):
        print("Network card disconnected")
        return True